src/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs

```diff
@@ -430,6 +469,34 @@ await workspace.TryOnDocumentOpenedAsync(
         }
     }
 
+    /// <summary>
+    /// Checks if the open source generator document contents matches the contents of the workspace solution.
+    /// This looks at the source generator state explicitly to avoid actually running source generators
+    /// </summary>
+    private static bool DoesAllSourceGeneratedTextMatchWorkspaceSolution(
+        ImmutableArray<(SourceGeneratedDocumentIdentity Identity, DateTime Generated, SourceText Text)> sourceGenereatedDocuments,
+        Solution workspaceSolution)
+    {
+        var compilationState = workspaceSolution.CompilationState;
+        foreach (var (identity, _, text) in sourceGenereatedDocuments)
+        {
+            var existingState = compilationState.TryGetSourceGeneratedDocumentStateForAlreadyGeneratedId(identity.DocumentId);
+            if (existingState is null)
+            {
+                // We don't have existing state for at least one of the documents, so the text does cannot match.
+                return false;
+            }
+
+            var newState = existingState.WithText(text);
```

Comment:
I forget, how expensive is this? If the state is different, can we actually use that later rather than doing it a second time?

Reply:
It doesn't seem to expensive (compares checksums, then copies over to a new Sg state.  As far as I can tell any actual work is deferred - https://source.dot.net/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/SourceGeneratedDocumentState.cs,c9182b28095a2b1c,references

if you'd prefer I can expose a version of `WithFrozenSourceGeneratedDocuments` that takes in the sg states.

Reply:
Looked into this a little bit.  Definitely possible to do with some small refactoring to `WithFrozenSourceGeneratedDocuments`.  However I don't think it will make much of a difference for the following reasons
1.  If the state is the same, the only work we do is calculating the checksums, which we would do regardless
2.  If the state is different, there still isn't much work done - we create a `SourceGeneratedDocumentState`, but all the parsing and everything else is deferred in lazy's.
3.  We create a max of one new state here if the text is different since this exists as soon as we detect one difference.

Think it is fine to leave it like this, but let me know what you think